### PR TITLE
Service Fabric Managed Clusters - API version 2024-06-01-preview

### PR DESCRIPTION
**Pull request by a-santamaria**
# ARM (Control Plane) API Specification Update Pull Request 

> [!TIP]
> Overwhelmed by all this guidance? See the `Getting help` section at the bottom of this PR description.

## PR review workflow diagram

Please understand this diagram before proceeding. It explains how to get your PR approved & merged.

spec_pr_review_workflow_diagram

## Purpose of this PR

What's the purpose of this PR? Check the specific option that applies. This is **mandatory**!

  - [ ] New resource provider.
  - [x] New API version for an existing resource provider. (If API spec is not defined in TypeSpec, the PR should have been created in adherence to [OpenAPI specs PR creation guidance](https://aka.ms/azsdkdocs/createopenapispec)).
  - [ ] Update existing version for a new feature. (This is applicable only when you are revising a private preview API version.)
  - [ ] Update existing version to fix OpenAPI spec quality issues in S360.
  - [ ] Other, please clarify:
    - _edit this with your clarification_

## Due diligence checklist

To merge this PR, you **must** go through the following checklist and confirm you understood 
and followed the instructions by checking all the boxes:

- [x] I confirm this PR is modifying Azure Resource Manager (ARM) related specifications, and not data plane related specifications.
- [x] I have reviewed following [Resource Provider guidelines](https://aka.ms/rpguidelines), including
  [ARM resource provider contract](https://aka.ms/azurerpc) and
  [REST guidelines](https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md) (estimated time: 4 hours).  
  I understand this is required before I can proceed to the diagram Step 2, "ARM API changes review", for this PR.

## Additional information

<details>
<summary> Viewing API changes</summary>

For convenient view of the API changes made by this PR, refer to the URLs provided in the table 
in the `Generated ApiView` comment added to this PR. You can use ApiView to show API versions diff. 

</details>
<details>
<summary>Suppressing failures</summary>

If one or multiple validation error/warning suppression(s) is detected in your PR, please follow the 
[suppressions guide](https://aka.ms/azsdk/pr-suppressions) to get approval.

</details>

## Getting help

- First, please carefully read through this PR description, from top to bottom. Please fill out the `Purpose of this PR` and `Due diligence checklist`.
- If you don't have permissions to remove or add labels to the PR, request `write access` per [aka.ms/azsdk/access#request-access-to-rest-api-or-sdk-repositories](https://aka.ms/azsdk/access#request-access-to-rest-api-or-sdk-repositories")
- To understand what you must do next to merge this PR, see the `Next Steps to Merge` comment. It will appear within few minutes of submitting this PR and will continue to be up-to-date with current PR state.
- For guidance on fixing this PR CI check failures, see the hyperlinks provided in given failure 
  and https://aka.ms/ci-fix.
- For help with ARM review (PR workflow diagram Step 2), see https://aka.ms/azsdk/pr-arm-review.
- If the PR CI checks appear to be stuck in `queued` state, please add a comment with contents `/azp run`.
  This should result in a new comment denoting a `PR validation pipeline` has started and the checks should be updated after few minutes.
- If the help provided by the previous points is not enough, post to https://aka.ms/azsdk/support/specreview-channel and link to this PR.

### CODE REVIEW

File: specification/servicefabricmanagedclusters/resource-manager/Microsoft.ServiceFabric/preview/2024-06-01-preview/nodetype.json
Comment (razvanbadea-msft) [inline specification/servicefabricmanagedclusters/resource-manager/Microsoft.ServiceFabric/preview/2024-06-01-preview/nodetype.json:1576]: Replace boolean/switch properties with a more meaningful enum whenever possible.
A boolean will forever have two valid values (true or false). A string enum type is always preferred. Also, properties should always provide better values just than True and False. For example two switches "isTypeA" and "isTypeB" should be replaced with one enum "type": [A, B, DefaultType]. 

Enums are always a more flexible and future proof option because they allow additional values to be added in the future in a non-breaking way, e.g. [Enabled, Disabled, Suspended, Deallocated].
Note: do NOT define a 'boolean enum' with two values 'True and False'. This might be easier to 'extend' in terms of types, but semantically its cofusing, and no better than a boolean.

File: specification/servicefabricmanagedclusters/resource-manager/Microsoft.ServiceFabric/preview/2024-06-01-preview/nodetype.json
Comment (razvanbadea-msft) [inline specification/servicefabricmanagedclusters/resource-manager/Microsoft.ServiceFabric/preview/2024-06-01-preview/nodetype.json:1547]: From the description, could you use the uri format?

File: specification/servicefabricmanagedclusters/resource-manager/Microsoft.ServiceFabric/preview/2024-06-01-preview/nodetype.json
Comment (a-santamaria) [reply specification/servicefabricmanagedclusters/resource-manager/Microsoft.ServiceFabric/preview/2024-06-01-preview/nodetype.json:1572]: its a single property, this definition is copied from vmss because its just a pass through to them: https://learn.microsoft.com/en-us/azure/templates/microsoft.compute/virtualmachinescalesets?tabs=bicep&pivots=deployment-language-bicep#vmgalleryapplication

File: specification/servicefabricmanagedclusters/resource-manager/Microsoft.ServiceFabric/preview/2024-06-01-preview/nodetype.json
Comment (ms-bogdan-rotaru) [reply specification/servicefabricmanagedclusters/resource-manager/Microsoft.ServiceFabric/preview/2024-06-01-preview/nodetype.json:1577]: Has this comment been addressed?

Comment (mentat9) [reply specification/servicefabricmanagedclusters/resource-manager/Microsoft.ServiceFabric/preview/2024-06-01-preview/nodetype.json:1577]: I don't see a response. @a-santamaria - you could consider something like deploymentCompletionPolicy: { AllowFailure, FailOnError }

File: specification/servicefabricmanagedclusters/resource-manager/Microsoft.ServiceFabric/preview/2024-06-01-preview/nodetype.json
Comment (mentat9) [inline specification/servicefabricmanagedclusters/resource-manager/Microsoft.ServiceFabric/preview/2024-06-01-preview/nodetype.json:1553]: Similar to earlier comment: please use enums instead of booleans. In this case, would something like upgradePolicy: { Automatic, Manual } work?

File: specification/servicefabricmanagedclusters/resource-manager/Microsoft.ServiceFabric/preview/2024-06-01-preview/nodetype.json
Comment (mentat9) [reply specification/servicefabricmanagedclusters/resource-manager/Microsoft.ServiceFabric/preview/2024-06-01-preview/nodetype.json:1573]: But your API is an ARM API, so you need to wrap the passthrough in an ARM/RPC-compatible way. You shouldn't use the name "tags" because it's confusing for clients, since it's nothing to do with resource tags in ARM. I'd recommend you rename this to vmGalleryTags or similar. You can still pass the value through.

**Comment by mentat9** [comment]
@a-santamaria - Please consider and respond to the active comments before resolving. In this case, please consider wrapping the VMGallery API with a more ARM-consistent API (per the feedback), rather than just straight pass-through.

**Comment by a-santamaria** [comment]
> @a-santamaria - Please consider and respond to the active comments before resolving. In this case, please consider wrapping the VMGallery API with a more ARM-consistent API (per the feedback), rather than just straight pass-through.

Ok, we have renamed the property tags to vmGalleryTags as suggested but kept the Boolean properties for the vmApps object

**Comment by ramoka178** [comment]
> > @a-santamaria - Please consider and respond to the active comments before resolving. In this case, please consider wrapping the VMGallery API with a more ARM-consistent API (per the feedback), rather than just straight pass-through.
> 
> Ok, we have renamed the property tags to vmGalleryTags as suggested but kept the Boolean properties for the vmApps object

Enums are highly suggested and is there a reason why the suggestions from previous reviewers do not work for you ?

**Comment by a-santamaria** [comment]
> > > @a-santamaria - Please consider and respond to the active comments before resolving. In this case, please consider wrapping the VMGallery API with a more ARM-consistent API (per the feedback), rather than just straight pass-through.
> > 
> > 
> > Ok, we have renamed the property tags to vmGalleryTags as suggested but kept the Boolean properties for the vmApps object
> 
> Enums are highly suggested and is there a reason why the suggestions from previous reviewers do not work for you ?

we decided to keep Booleans because its the same definition as vmss, so they are not going to change the behavior and that way customers can use the same definitions the have for vmss in sfmc

**Comment by gary-x-li** [comment]
Please resolve merging conflict and add the label `PublishToCustomers` to confirm that you understand the spec will be published to customer.

**Comment by mentat9** [comment]
> @a-santamaria - Your PR is in the merge queue, but we aren't able to merge it until you fix the merge conflicts.

@a-santamaria - Just add back the MergeRequested label after resolving the merge conflicts if you still need this PR merged.

**Comment by a-santamaria** [comment]
> > @a-santamaria - Your PR is in the merge queue, but we aren't able to merge it until you fix the merge conflicts.
> 
> @a-santamaria - Just add back the MergeRequested label after resolving the merge conflicts if you still need this PR merged.

fixed the conflict by rebasing and added back MergeRequested label, thanks!

**Review by mentat9 — APPROVED** [review]

**Comment by mentat9** [comment]
> > > @a-santamaria - Your PR is in the merge queue, but we aren't able to merge it until you fix the merge conflicts.
> > 
> > 
> > @a-santamaria - Just add back the MergeRequested label after resolving the merge conflicts if you still need this PR merged.
> 
> fixed the conflict by rebasing and added back MergeRequested label, thanks!

@a-santamaria - All required checks need to pass. You still have one required check failing: https://github.com/Azure/azure-rest-api-specs/pull/29327/checks?check_run_id=29015767632. Add back the MergeRequested label after fixing all required checks are passing.

**Comment by a-santamaria** [comment]
> > > > @a-santamaria - Your PR is in the merge queue, but we aren't able to merge it until you fix the merge conflicts.
> > > 
> > > 
> > > @a-santamaria - Just add back the MergeRequested label after resolving the merge conflicts if you still need this PR merged.
> > 
> > 
> > fixed the conflict by rebasing and added back MergeRequested label, thanks!
> 
> @a-santamaria - All required checks need to pass. You still have one required check failing: https://github.com/Azure/azure-rest-api-specs/pull/29327/checks?check_run_id=29015767632. Add back the MergeRequested label after fixing all required checks are passing.

@mentat9 the requried check was just complaing that I had to add PublishToCustomers label I don't see MergeRequested anymore

**Review by mentat9 — APPROVED** [review]

**Comment by mentat9** [comment]
@a-santamaria - The merge process changed, and I didn't get the memo. PR authors with write access to the repo now can merge their own PRs when all required checks have passed. There is no longer a step to request merge.

I would have merged your PR for you right now, but there was a stuck check preventing it, so I reran the pipeline. I will revisit your PR after the checks finish running and do the merge for you to help expediate things. Or if you get here before I do, feel free to merge it yourself. Sorry for giving you outdated instructions previously.

**Comment by mentat9** [comment]
> @a-santamaria - The merge process changed, and I didn't get the memo. PR authors with write access to the repo now can merge their own PRs when all required checks have passed. There is no longer a step to request merge.
> 
> I would have merged your PR for you right now, but there was a stuck check preventing it, so I reran the pipeline. I will revisit your PR after the checks finish running and do the merge for you to help expediate things. Or if you get here before I do, feel free to merge it yourself. Sorry for giving you outdated instructions previously.

@a-santamaria - The same check (Protected Files) is still not passing. I don't know anything about this. I'd recommend asking for help from the experts on the Teams channel: https://teams.microsoft.com/l/channel/19%3A0351f5f9404446e4b4fd4eaf2c27448d%40thread.skype/API%20Spec%20Review?groupId=3e17dcb0-4257-4a30-b843-77f47f1d4121&tenantId=72f988bf-86f1-41af-91ab-2d7cd011db47.

**Pull request merged**
